### PR TITLE
[Fix] Fix reduction=mean in CELoss.

### LINKED ISSUE #7367: Inconsistency with loss function of PyTorch when using ingore_index arg
Hello，I'm a user of mmdet3d, as this toolbox import a lot of funcs from mmdet, which is the reason why I report here.

For example the commly used module  `CrossEntropyLoss` with `ignore_index`

Here is the definition from PyTorch

```
        ignore_index (int, optional): Specifies a target value that is ignored
            and does not contribute to the input gradient. When :attr:`size_average` is
            ``True``, the loss is averaged over non-ignored targets.
```

And the following is from MMDetection

```
        ignore_index (int | None): The label index to be ignored.
            If None, it will be set to default value. Default: -100.
```

This is the [code segment](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/losses/cross_entropy_loss.py) from mmdet 

```python
def cross_entropy(pred,
                  label,
                  weight=None,
                  reduction='mean',
                  avg_factor=None,
                  class_weight=None,
                  ignore_index=-100):
    """Calculate the CrossEntropy loss.

    Args:
        pred (torch.Tensor): The prediction with shape (N, C), C is the number
            of classes.
        label (torch.Tensor): The learning label of the prediction.
        weight (torch.Tensor, optional): Sample-wise loss weight.
        reduction (str, optional): The method used to reduce the loss.
        avg_factor (int, optional): Average factor that is used to average
            the loss. Defaults to None.
        class_weight (list[float], optional): The weight for each class.
        ignore_index (int | None): The label index to be ignored.
            If None, it will be set to default value. Default: -100.

    Returns:
        torch.Tensor: The calculated loss
    """
    # The default value of ignore_index is the same as F.cross_entropy
    ignore_index = -100 if ignore_index is None else ignore_index
    # element-wise losses
    loss = F.cross_entropy(
        pred,
        label,
        weight=class_weight,
        reduction='none',
        ignore_index=ignore_index)

    # apply weights and do the reduction
    if weight is not None:
        weight = weight.float()
    loss = weight_reduce_loss(
        loss, weight=weight, reduction=reduction, avg_factor=avg_factor)

    return loss
```

The key is that, the results are inconsistent. If someone uses mmdets `CrossEntropyLoss` by default and give a param of `ignore_index` such as 255, torch version will ignores the ignored elements, while mmdet version will first caclulate element-wise loss, then apply `mean` reduction for all elements by default. 

In my scenario of point cloud segmentation, the ground truth is a voxelized tensor with most of the labels ignored. By torch version, `mean` reduction will be performed over only no-ingored voxels while mmdet version over all elements.

It troubles me for days  because of this tiny, sort of, 'bugs'?

Pls figure it out and keep consistency with common funcs of torch

Bests

Singal

### PR DESCRIPTION
**Pull request by RangiLyu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

resolve #7367 

## Modification

Fix avg_factor in CELoss with ignore index.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -41,6 +41,9 @@ def cross_entropy(pred,
         reduction='none',
         ignore_index=ignore_index)
 
+    if reduction == 'mean' and avg_factor is None:
```
Comment (ZwwWayne) [inline mmdet/models/losses/cross_entropy_loss.py:44]: Add comments to explain the reason. It is not straightforward to understand that it aims to calculate the real mean value of the loss.

File: mmdet/models/losses/cross_entropy_loss.py [hunk 2]
```diff
@@ -41,6 +41,10 @@ def cross_entropy(pred,
         reduction='none',
         ignore_index=ignore_index)
 
+    # loss is averaged over non-ignored targets
+    if reduction == 'mean' and avg_factor is None:
+        avg_factor = max(1, (label != ignore_index).sum())
```
Comment (hhaAndroid) [inline mmdet/models/losses/cross_entropy_loss.py:46]: Why not `avg_factor = max(1, ((label >= 0) & (label != ignore_index)).sum())`

File: mmdet/models/losses/cross_entropy_loss.py (hunk 1 above)
Comment (hhaAndroid) [reply mmdet/models/losses/cross_entropy_loss.py:45]: Post-related issues.

**Review by ZwwWayne — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]
